# 错误处理

## .parse()

+ 当验证失败时，.parse() 方法将抛出一个 ZodError 实例，其中包含有关验证问题的详细信息

  ```js
  try {
    Player.parse({ username: 42, xp: "100" });
  } catch(error){
    if(error instanceof z.ZodError){
      error.issues;
      /* [
        {
          expected: 'string',
          code: 'invalid_type',
          path: [ 'username' ],
          message: 'Invalid input: expected string'
        },
        {
          expected: 'number',
          code: 'invalid_type',
          path: [ 'xp' ],
          message: 'Invalid input: expected number'
        }
      ] */
    }
  }
  ```

## safeParse

+ 要避免 try/catch 阻塞，你可以使用 `.safeParse()` 方法返回包含成功解析数据或 ZodError 的纯文本结果对象
+ 结果类型为 可区分联合，因此你可以方便地处理这两种情况

  ```js
  const result = Player.safeParse({ username: 42, xp: "100" });
  if (!result.success) {
    result.error;   // ZodError instance
  } else {
    result.data;    // { username: string; xp: number }
  }
  ```

## .safeParseAsync()

+ 异步 API，例如 async、refinements 或 transforms，则需要改用 .safeParseAsync() 方法。

  ```js
  await schema.safeParseAsync("hello");
  ```
